### PR TITLE
Fix connecting user from background thread

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🔗 Issue Links
https://github.com/GetStream/stream-chat-swift/issues/2702

### 🎯 Goal
Make sure it is possible to connect a user from a background thread. Previously, this would cause a crash, or nothing would happen, and the connection would never finish.

### 🛠 Implementation
The root cause of the issue is that every dependency of `ChatClient` was lazily initialised. Since lazy properties are not thread-safe, this would cause unpredictable behaviour. Besides that, there were circular dependencies too.

### 🎨 Showcase

| Before | After |
| ------ | ----- |
| <video src="https://github.com/GetStream/stream-chat-swift/assets/12814114/4970989b-0ec4-464d-80ae-467c24e93942"/> |  <video src="https://github.com/GetStream/stream-chat-swift/assets/12814114/843c772a-47c3-426d-acbf-208afe00ac33"/> |

### 🧪 Manual Testing Notes
1. Connect the User with a Demo App
2. Should connect successfully

We now always connect from a background thread in the DemoApp to make sure we cover this scenario.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 1]
```diff
@@ -16,10 +16,10 @@ class APIClient {
     let decoder: RequestDecoder
 
     /// Used for reobtaining tokens when they expire and API client receives token expiration error
-    let tokenRefresher: (@escaping () -> Void) -> Void
+    var tokenRefresher: ((@escaping () -> Void) -> Void)?
```
Comment (martinmitrevski) [inline Sources/StreamChat/APIClient/APIClient.swift:19]: why do we make these optional?

File: Sources/StreamChat/APIClient/RequestEncoder.swift [hunk 2]
```diff
@@ -70,7 +70,7 @@ extension RequestEncoder {
 }
 
 /// The default implementation of `RequestEncoder`.
-struct DefaultRequestEncoder: RequestEncoder {
+class DefaultRequestEncoder: RequestEncoder {
```
Comment (martinmitrevski) [inline Sources/StreamChat/APIClient/RequestEncoder.swift:73]: Why do we change it to a class?

File: Sources/StreamChat/ChatClient.swift [hunk 3]
```diff
@@ -49,206 +49,48 @@ public class ChatClient {
...
+    // MARK: Repositories
 
-        center.add(middlewares: middlewares)
+    private(set) var connectionRepository: ConnectionRepository
 
-        return center
-    }()
-
-    // MARK: Repositories
+    private(set) var authenticationRepository: AuthenticationRepository
```
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:58]: Will these change at some point? Maybe `let` would be better?

File: Sources/StreamChat/ChatClient.swift [hunk 4]
```diff
@@ -49,206 +49,48 @@ public class ChatClient {
...
-                    completion: { _ in completion() }
-                )
-            },
-            { [weak self] endpoint in
-                self?.syncRepository.queueOfflineRequest(endpoint: endpoint)
-            }
-        )
-        return apiClient
-    }()
+    var apiClient: APIClient
```
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:77]: this one is both `var` and the setter is not private.

File: Sources/StreamChat/ChatClient.swift [hunk 5]
```diff
@@ -285,6 +127,94 @@ public class ChatClient {
         self.config = config
         self.environment = environment
 
+        let configuration = config.urlSessionConfiguration
```
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:130]: maybe a separate setup method for readability?

File: Sources/StreamChat/ChatClient.swift [hunk 6]
```diff
@@ -285,6 +127,94 @@ public class ChatClient {
...
+        self.syncRepository = syncRepository
+        authenticationRepository = authRepository
+        extensionLifecycle = environment.extensionLifecycleBuilder(config.applicationGroupIdentifier)
+        callRepository = environment.callRepositoryBuilder(apiClient)
+        channelRepository = environment.channelRepositoryBuilder(
+            databaseContainer,
+            apiClient
+        )
+
+        authRepository.delegate = self
```
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:204]: The additional setup delegate stuff could be also moved into separate methods, since this one is becoming too big.

File: Sources/StreamChat/ChatClient.swift [hunk 7]
```diff
@@ -516,24 +446,197 @@ extension ChatClient: ConnectionDetailsProviderDelegate {
...
+            """
+                Missing token refresh provider to get a new token
+                When using expiring tokens you need to provide a way to refresh it by passing `tokenProvider` when \
+                calling `ChatClient.connectUser()`.
+            """
+        }
+    }
+}
+
+class ChatClientFactory {
```
Comment (martinmitrevski) [inline Sources/StreamChat/ChatClient.swift:504]: This is nice 👍 Maybe move it in a separate file?

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift [hunk 8]
```diff
@@ -14,8 +14,6 @@ class TypingStartCleanupMiddleware: EventMiddleware {
     /// A closure that will be invoked with `stop typing` event when the `incomingTypingStartEventTimeout` has passed
     /// after `start typing` event.
     let emitEvent: (Event) -> Void
-    /// A closure to get a list of user ids to skip typing events for them.
-    let excludedUserIds: () -> Set<UserId>
```
Comment (martinmitrevski) [inline Sources/StreamChat/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift:18]: Why do we remove this?

**Review by martinmitrevski — COMMENTED** [review]
Good that we tackle the lazy vars 👍  We should revisit few things:
- some properties are `var` and others `let`, we should have some consistency (preferably all are constants setup on `init`)
- optionality is introduced in few places, do we need that? 
- we should split up the init method, tends to be hard to follow

File: Sources/StreamChat/APIClient/RequestEncoder.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChat/APIClient/RequestEncoder.swift:73]: We need the reference semantics. Because not this is created in the init without keeping any reference to the encoder. From what I remember, all requests were failing if we kept them as structs.

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift (hunk 8 above)
Comment (nuno-vieira) [reply Sources/StreamChat/WebSocketClient/EventMiddlewares/TypingStartCleanupMiddleware.swift:18]: This was on of the responsible for the circular dependency. We were using the `authenticationRepository` to get the current user, and populate here the `excludedUserIds`. This is over complicated for something that is really simple todo. We have access to the current user in the `DatabaseContainer` inside this `Middleware`. So we have simplified the code, and fixed the circular dependency 👍

File: Sources/StreamChat/ChatClient.swift (hunk 3 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift:58]: Yes this can be all `let` 👍 They were `var` becasue they were lazy before, now we can change them to `let`. Will do 👍

File: Sources/StreamChat/ChatClient.swift (hunk 4 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift:77]: The setter was not private before, I will need to investigate if this is user for testing or not 🤔

File: Sources/StreamChat/ChatClient.swift (hunk 7 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift:504]: Yes I can move it to a separate file 👍

File: Sources/StreamChat/ChatClient.swift (hunk 6 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift:204]: I think it is useful that everything is in the same place. If I move this to a different method, I will also need to pass all the objects required for the setup 🤔 I will see if I can improve this either way

**Comment by nuno-vieira** [comment]
> Good that we tackle the lazy vars 👍 We should revisit few things:
> 
> * some properties are `var` and others `let`, we should have some consistency (preferably all are constants setup on `init`)
> * optionality is introduced in few places, do we need that?
> * we should split up the init method, tends to be hard to follow

> optionality is introduced in few places, do we need that?

Do you mean some closures? Yes we need those because we need to populate them in the init of `ChatClient` and this implementations require the reference of ` ChatClient`, so we need to initialise all the dependencies first, and only then we can populate this closures

File: Sources/StreamChat/APIClient/APIClient.swift (hunk 1 above)
Comment (nuno-vieira) [reply Sources/StreamChat/APIClient/APIClient.swift:19]: The explaination is here: https://github.com/GetStream/stream-chat-swift/pull/2762#issuecomment-1703066227

File: Sources/StreamChat/ChatClient.swift (hunk 7 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift]: ✅ I've also moved the Environemnt struct to a new file as well, to make it consistent.

File: Sources/StreamChat/ChatClient.swift (hunk 6 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift:201]: Improved these ones a bit 👍

File: Sources/StreamChat/ChatClient.swift (hunk 5 above)
Comment (nuno-vieira) [reply Sources/StreamChat/ChatClient.swift:128]: This one was not possible to extract to a separate function because it needs to be configured before passing it to the `APIClient`. At this time `self` is not available yet.
